### PR TITLE
Mobilenet

### PR DESCRIPTION
**Pull request by dwSun**
## Description ##
mobilenet:
add hyper parameters alpha and resolution
add extra compact symbol with less lines of code

### Essentials ###
- [ y ] Passed code style checking (`make lint`)
- [ y ] Changes are complete (i.e. I finished coding on this PR)
- [ y ] All changes have test coverage
- [ y ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ y ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
mobilenet:
add hyper parameters alpha and resolution
add extra compact symbol with less lines of code

## Comments ##
- detail symbol implementation maybe be useful for someone who want to implement ssd-mobilenet

### CODE REVIEW

**Comment by dwSun** [comment]
So last pull request not merged will show here too.

File: example/image-classification/symbols/mobilenet.py [hunk 1]
```diff
@@ -14,48 +14,129 @@
...
-
-    pool = mx.sym.Pooling(data=conv_14, kernel=(7, 7), stride=(1, 1), pool_type="avg", name="global_pool")
+
+def Conv_DPW(data, depth=1, stride=(1, 1), name='', idx=0, suffix=''):
+    conv_dw = Conv(data, num_group=depth, num_filter=depth, kernel=(3, 3), pad=(1, 1), stride=stride, name="conv_%d_dw" % (idx), suffix=suffix)
+    conv = Conv(conv_dw, num_filter=depth * stride[0], kernel=(1, 1), pad=(0, 0), stride=(1, 1), name="conv_%d" % (idx), suffix=suffix)
+    return conv
+
+
+def get_symbol_compact(num_classes, alpha=1, resolution=224, **kwargs):
```
Comment (zhreshold) [inline example/image-classification/symbols/mobilenet.py]: This function name will not be recognized by train_imagenet.py

File: example/image-classification/symbols/mobilenet.py [hunk 2]
```diff
@@ -14,48 +14,129 @@
...
+
+    pool_size = int(resolution / 32)
+    pool = mx.sym.Pooling(data=conv_14_dpw, kernel=(pool_size, pool_size), stride=(1, 1), pool_type="avg", name="global_pool")
+    flatten = mx.sym.Flatten(data=pool, name="flatten")
+    fc = mx.symbol.FullyConnected(data=flatten, num_hidden=num_classes, name='fc')
+    softmax = mx.symbol.SoftmaxOutput(data=fc, name='softmax')
+    return softmax
+
+
+def get_symbol(num_classes, alpha=1, resolution=224, **kwargs):
```
Comment (dwSun) [inline example/image-classification/symbols/mobilenet.py]: This is the function prepared for train_imagenet.py

File: example/image-classification/symbols/mobilenet.py (hunk 1 above)
Comment (dwSun) [reply example/image-classification/symbols/mobilenet.py]: This function is a compact version of **get_symbol** with less lines of code. The network defined in this function is identical with **get_symbol**.
I should add more docstring to explain that.

File: example/image-classification/symbols/mobilenet.py (hunk 2 above)
Comment (zhreshold) [reply example/image-classification/symbols/mobilenet.py]: hmm. I don't get the point of providing two identical function with one more compact than the other. I think it's good enough to simply provide the compact one as long as the internal names are unchanged.

Comment (dwSun) [reply example/image-classification/symbols/mobilenet.py]: Someone may need details about the network implementation.

**Comment by piiswrong** [comment]
@zhreshold Could you follow up?

**Comment by zhreshold** [comment]
@dwSun One more qustion, is this change fully compatible with previously trained model?

**Comment by dwSun** [comment]
With alpha=1, this change is fully compatible with previously trained model.

The symbol loaded by mx.model.load_checkpoint is slightly different with the one return by mobilenet.get_symbol. The global_pool kernel size is calculated depending on resolution in this change.

**Comment by zhreshold** [comment]
@dwSun Names are inconsistent.
The original param names are 
```
'conv12_pointwise_bias', 'batchnorm1_gamma', 'conv12_pointwise_weight', 'conv7_depthwise_bias'
```
Now they are
```
'conv_9_dw_conv2d_weight', 'conv_1_batchnorm_beta', 'conv_8_dw_conv2d_weight'
```
Which means old models are no longer usable after this change. Can you follow the old names to not break the backward compatibility?
Other changes looks good.

**Comment by dwSun** [comment]
@zhreshold, I checked the symbol, with:
```py
import mobilenet_mine  # my PR version
import mobilenet_origin  # 41394a448a92efbe7704fe21c48a8504fd8d3edc,example/image-classification/symbols/mobilenet.py

sym_orig = mobilenet_origin.get_symbol(10)
sym_mine = mobilenet_mine.get_symbol(10)
sym_mine_compcat = mobilenet_mine.get_symbol_compact(10)

with open('sym_orig.json', mode='w') as of:
    of.write(sym_orig.tojson())

with open('sym_mine.json', mode='w') as of:
    of.write(sym_mine.tojson())

with open('sym_mine_compcat.json', mode='w') as of:
    of.write(sym_mine_compcat.tojson())
```
Those json files are identical to each other.

> Which means old models are no longer usable after this change.
Are those old models from other symbol definition?

File: example/image-classification/symbols/mobilenet.py [hunk 3]
```diff
@@ -14,48 +14,129 @@
...
 # under the License.
-
+# -*- coding:utf-8 -*-
 import mxnet as mx
 
-def Conv(data, num_filter=1, kernel=(1, 1), stride=(1, 1), pad=(0, 0), num_group=1, name=None, suffix=''):
-    conv = mx.sym.Convolution(data=data, num_filter=num_filter, kernel=kernel, num_group=num_group, stride=stride, pad=pad, no_bias=True, name='%s%s_conv2d' %(name, suffix))
-    bn = mx.sym.BatchNorm(data=conv, name='%s%s_batchnorm' %(name, suffix), fix_gamma=True)
-    act = mx.sym.Activation(data=bn, act_type='relu', name='%s%s_relu' %(name, suffix))
+__author__ = 'qingzhouzhen'
```
Comment (zhreshold) [inline example/image-classification/symbols/mobilenet.py]: Please use standard __authors__, __date__, and place before import per PEP8 standard.

**Comment by zhreshold** [comment]
Okay, it's my bad checking the wrong version of mobilenet symbol. 
Please change the dunder names and rebase to fix the CI so we can merge it.
Thanks for the contribution.

**Comment by dwSun** [comment]
sorry I deleted the remote branch mistakenly. I have to start a new PR here #9099.

**Pull request closed without merging**